# 🐳 Snapshots Testing

## Snapshot Test

Snapshot Test는 서로 다른 두 출력을 비교해 어설션 하는 기술.

수행되는 방식은 종단 간 테스트의 [시각적 회귀 테스트](https://ideveloper2.dev/blog/2021-01-24--%EC%8B%9C%EA%B0%81%EC%A0%81-%ED%9A%8C%EA%B7%80-%ED%85%8C%EC%8A%A4%ED%8A%B8-visual-regression-test/)와 유사



**Example**

버튼의 스타일이 변경되지 않았는지 테스트

버튼의 기본 스크린샷을 찍고, 테스트 도구 모음이 실행될때마다 새 스크린샷을 찍고 기본 스크린샷과 픽셀단위로 비교함.



스냅샷 테스트는 동일한 기술을 적용하지만 이미지 대신 텍스트 또는 json과 같은 직렬화 가능한 출력을 기반으로 함.

**매개변수를 기반으로** 오류 메세지를 반환하고 조건과 일치하지 않는 경우 오류를 발생시키는 함수를 살펴보자

```javascript
// error.js
export default function getErrorMessage(code) {
  if (code === 1) {
    return "The camel walks on a leg";
  } else if (code === 2) {
    return "Rabbits don't eat carrots";
  } else if (code === 3) {
    return "Cats don't eat mouses";
  }
  throw new Error("No error messages for that code");
}
```

지금 까지 배운걸로 시도하면

```javascript
import getErrorMessage from "./error-message";

describe("getErrorMessage", () => {
  it("returns camel message when code is 1", () => {
    expect(getErrorMessage(1)).toBe("The camel walks on a leg");
  });

  it("returns rabbit message when code is 2", () => {
    expect(getErrorMessage(2)).toBe("Rabbits don't eat carrots");
  });

  it("returns cat message when code is 3", () => {
    expect(getErrorMessage(3)).toBe("Cats don't eat mouses");
  });

  it("throws an error otherwise", () => {
    expect(() => getErrorMessage(4)).toThrow("No error messages for that code");
  });
});
```



문제점을 살펴보자.

* **반복적**이다. => 동일한 논리로 각 사례에 대한 테스트가 있다.
* 소스 코드측에서 메시지를 변경하면 테스트를 변경해야함.<br/>❗ 자주 업데이트 해야하는 테스트는 좋은 테스트가 아니다.

스냅샷 테스트를 사용해 테스트에 동일한 값을 제공하면서 이런 문제를 피할 수 있다.



직렬화 가능한 값을 위한 `toMatchSnapshot`

발생한 오류를 위한 `toThrowErrorMatchin`

```javascript
import getErrorMessage from "./error-message";

describe("getErrorMessage", () => {
  it("returns an error for a valid code", () => {
    expect(getErrorMessage(1)).toMatchSnapshot();
    expect(getErrorMessage(2)).toMatchSnapshot();
    expect(getErrorMessage(3)).toMatchSnapshot();
  });

  it("throws an error otherwise", () => {
    expect(() => getErrorMessage(4)).toThrowErrorMatchingSnapshot();
  });
});
```



getErrorMessage함수에서 코드가1일때 반환되는 값을 변경해보자.

```javascript
if (code === 1) {
  return "The dog walks on a leg";
}
```

스냅샷 테스트가 실패했음을 알리는 에러가 발생

> As you can see, it’s giving us the specific error, even pointing to the line of the `expect` statement that doesn’t match the snapshot.****

두 가지 경우로 생각해 볼 수 있다.

* 실수를 했고, 테스트에서 실제 오류가 발생
* 코드를 의도적으로 변경했으므로 스냅샷이 오래되었으며 업데이트 해야 함.



## 환경 구성

```shell
npm install --save-dev jest-serializer-vue
```

```json
// package.json
{
  // ...
  "jest": {
    // ...
    // serializer for snapshots
    "snapshotSerializers": [
      "jest-serializer-vue"
    ]
  }
}
```

```javascript
import { mount } from '@vue/test-utils';
import HelloWorld from '../HelloWorld.vue';

describe('Hello World Component', () => {
  test('[Snapshot Testing] renders hello world message', () => {
    const { vm } = mount(HelloWorld);
    expect(vm.$el).toMatchSnapshot();
  });
});
 
```

❗ 명령어를 `npm t -u`로 해야함.